### ISSUE TITLE
PyUpdater Packaged app does not restart on mac

### ISSUE THREAD

**Issue by JessicaTegner** [Digital-Sapphire/PyUpdater#207]
Hi.

I'm using PyUpdater to provide updates for my applications both on windows and on mac, but when my code, applies an update on the mac side, it just goes busy and hang, but on windows it works fine and restarts and everything.

### System information:
Systems that my code are tested on.
* Windows 10 1903.
* Mac OS Mojave.
Python stuff:
* Python version: 3.7-32 bit (64 bit on mac)
* PyUpdater version: "3.1.1+3.gbed2239"
* Update channels tested: "Alpha and Beta"
* Pyupdater notes: Patching disable (shouldn't say anything, because I'm not using stable atm).
* Application UI Toolkit: WXPython version 4.0.3.

### Code example:
^^
	updater_client = Client(ClientConfig())
	updater_client.refresh()
	progress_bar = wx.ProgressDialog(_("VisCon Updater"), _("Downloading update. Please wait!"), 100, None, wx.PD_APP_MODAL|wx.PD_AUTO_HIDE|wx.PD_ELAPSED_TIME|wx.PD_ESTIMATED_TIME|wx.PD_REMAINING_TIME)
	updater_client.add_progress_hook(lambda info: update_bar(info, progress_bar))
	app_update = updater_client.update_check(variables.app_name, variables.version, variables.release_channel)
	if app_update is not None:
		variables.update_in_progress = True
		logging.debug("Update found.")
		logging.info("Updating VisCon.")
		logging.debug("Cleaning up old updates.")
		app_update.cleanup()
		logging.debug("cleanup done.")
		t = threading.Thread(daemon=True, target=download_update, args=(app_update, progress_bar,))
		t.start()
		progress_bar.ShowModal()

def update_bar(info, bar):
	x = info.get(u'downloaded')/info.get(u'total')*100
	wx.CallAfter(bar.Update, x)

def download_update(app_update, dialog):
	logging.info("Downloading update-package.")
	app_update.download(background=True)
	while app_update.is_downloaded() == False:
		pass
	logging.debug("Download done, removing progress bar.")
	wx.CallAfter(dialog.Destroy)
	logging.info("Preparing for update.")
	wx.CallAfter(wx.GetTopLevelWindows()[0].Close)
	logging.info("Executing update.")
	app_update.extract_restart()
``

Looking at #30 I have also tried to change the last function to:
``
def download_update(app_update, dialog):
	logging.info("Downloading update-package.")
	app_update.download(background=True)
	while app_update.is_downloaded() == False:
		pass
	logging.debug("Download done, removing progress bar.")
	wx.CallAfter(dialog.Destroy)
	logging.info("Preparing for update.")
	wx.CallAfter(wx.GetTopLevelWindows()[0].Close)
	logging.info("Executing update.")
	if platform.system() == "Windows":
		app_update.extract_restart()
		return
	else:
		app_update.extract_overwrite()
		subprocess.Popen(sys.argv[0])
		sys.exit()

``

Without success (it still works fine on windows as expected).

### expected behavior
The user presses yes to download and install the update (not included here for brevity), the progress bar pops up, and shows the progress.
When the download is complete, the progress bare gets destroyed, the window(s) of the program closes and the update is applied and the program restarts (update extracted followed by program restart).

### What's actually happening:
On windows it goes like described above.
On mac the following happens instead.
* User says yes to download and install the update.
* The progress bar pops up and shows the download progress.
* At some point near completion, the progress bar disappears and then nothing happens.
* After around 2-5 seconds, the program goes busy and stops responding.
* After a while, the standard apple "this program unexpectedly stopped responding" dialog pops up and asks if you want to terminate it.
* After force quitting the program and launching it again, the update hasn't been applied and the dialog that asks the user if they want to update ops up again.
* If the user presses yes, the progress bar shows for a split second, and then it's a 50/50 chance that it does the same described above, or actually applies the update and restarts.

### Log output:
No errors in the log output, even with unexpected errors catched with sys.excepthook.

Hope I get a fix soon (I'm going mad lol :P )

**Comment by JMSwag** [comment]
Hey @NicklasMCHD, try the example below and let me know how that goes.

```python
def download_update(app_update, dialog):
    logging.info("Downloading update-package.")
    app_update.download(background=True)
    
    while app_update.is_downloaded() == False:
        pass
    
    logging.debug("Download done, removing progress bar.")
    wx.CallAfter(dialog.Destroy)
    
    logging.info("Preparing for update.")
    wx.CallAfter(wx.GetTopLevelWindows()[0].Close)
    
    logging.info("Executing update.")
    app_update.extract_restart()
```

**Comment by JessicaTegner** [comment]
Hi @JMSwag 
I don't see the difference between my first code snippet and your code snippet. Mind telling me the difference?

**Comment by JMSwag** [comment]
I didn’t mean to include the line that closes the top level window.

**Comment by JessicaTegner** [comment]
Ok @JMSwag 
I have tried to insert the code you wrote, and now Windows stopped working (haven't tested on mac yet).
* It extracts the update to the folder "C:\Users\Nickl\AppData\Local\COMPANY_NAME\APPLICATION_NAME\", like it should, the main program also shuts down, but it doesn't restart, and if you start it again manually, it still uses the old version (it asks again, if you want to update) and if you press yes, it just immediately closes and doesn't restart (because the update already is downloaded).
I need to have that main window close call there, because the user needs to get a chance to save his/her stuff before the program closes and executes the update.

**Comment by JMSwag** [comment]
Would you be able to provide a minimal script that reproduces this issue?

**Comment by JessicaTegner** [comment]
Sure thing. I have attached a sample script.

My setup with the repo was the following:
* Application name: calculator
* Your / company name: Accessiware
* Allow patches: n

To reproduce my setup do the following.
* install the requirements from the first comment (pyupdater from git and wxpython==4.0.3, i also used the scp plugin to upload).
* Follow the guide from https://github.com/Digital-Sapphire/PyUpdater/blob/master/docs/usage-cli.md
* The script I attached is the implementation.

What I ran after setting up the repo and my scp credentials:
> pyupdater make-spec --onedir --console calculator.py
> pyupdater build --app-version=1.0.0a1 win.spec
> pyupdater pkg -p -s --split-version
> pyupdater upload -s scp

Then after upload, I changed the version number from "1.0.0a1" to "1.0.0a2" and then created a new update package (using the same 3 last commands above, just with the new version number).

[calculator.py.txt](https://github.com/Digital-Sapphire/PyUpdater/files/3360690/calculator.py.txt)
[calculator.py.txt](https://github.com/Digital-Sapphire/PyUpdater/files/3360695/calculator.py.txt)
And same result as in comment 1.

**Comment by JessicaTegner** [comment]
Upsi: i accidentally uploaded the test script twice. It's the same script, down to the byte :) Sorry about that :)

**Comment by JMSwag** [comment]
Thanks, let me poke around and see what I can find.

**Comment by JessicaTegner** [comment]
hi @JMSwag 
Any updates?

**Comment by JMSwag** [comment]
Not at the moment. I will update this issue once I have something.

**Comment by JessicaTegner** [comment]
Ok just thought I would update this issue with my findings.
After it downloads the update and proceed to extract it and close, nothings happens as said before, so I went to where PyUpdater extracts the update and here's what I found.

* The downloaded file is not corrupted (hashes match with the one on the server)
* However, when it get's extracted something funny happens.

The application (version 1.0.0a1) is 29.9MB and the extracted version (1.0.0a2) in the update folder (on my system "C:\Users\USERNAME\AppData\Local\COMPANY_NAME\APP_NAME\") is only 7,73MB, and I go into the folder, there's only 12 rows of files and folders versus if extract the downloaded zip manually, which in that case, there's 38 rows of files and folders, the same as in the original (v 1.0.0a1) application, in addition to the size also beeing closer to the original,  at 29.9MB.

So in short: It seams like, that there's something strange going on during the extraction of the downloaded update.

Hope you can use this information to solve the issue @JMSwag

**Comment by JMSwag** [comment]
@NicklasMCHD Thanks for the updated info. Are you using folder instead of one file? If so, try with one file, if possible, and let me know how that goes. There have been a couple of reports with one-folder builds having updating issues.

**Comment by JessicaTegner** [comment]
Yah @JMSwag  I was running and testing with --onedir.
Now after trying with --onefile, it does even less, the update gets downloaded but it doesn't even get extracted.
And the program just closes and nothing happens.
* I'm using the script I sent before.

**Comment by JessicaTegner** [comment]
Ok, I just took another look at PyUpdaters source code.
To make things easyer to debug, it would be nice, if the output of the "robocopy" and the "move" commands from the .bat scripts, they could be stored in something like self.update_folder/logs or something, and be deleted on app_update.clean() (or wherever the clean function is located)

**Comment by webpolis** [comment]
Same thing happens on Windows. All the hashes and file sizes are ok in the versions.gz file. The PyUpdater downloads, but then nothing happens. I use `extract_restart` and it doesn't restarts. If I go to the _C:\Users\USERNAME\AppData\Local\COMPANY_NAME\APP_NAME_ folder where PyUpdater temporarily stores meta data and downloads files, I see the file shows with a `.part` extension, not `.zip`. The progress states its _Finished_.

pyu.log shows nothing relevant

**Comment by EvgenKoshevoy** [comment]
Hello. I have the same issue. If i load '.app' file, my app won't update, but if i load binary from /Contents/MacOS folder - everything is going well.

macOS 10.15.2
Python 3.7.3

    def update_app(self):
        self.app_update.cleanup()
        self.app_update.download()
        if self.app_update.is_downloaded():
            self.app_update.extract_restart()

**Comment by JMSwag** [comment]
I'm unable to reproduce this issue. A PR with a fix is welcome.

**Comment by JessicaTegner** [comment]
Hi.
@JMSwag  and others, I think I've discovered the problem.
On windows everything works, because the apps gets relaunched (you download the update, then call restart and then make your current instance quit).
On mac os (or any non-win platform) PyUpdater uses os.execl instead of a .bat (or on mac a .cmd) script to perform the restart.
The problem with os.execl is, that it doesn't actually start a new instance of the application, but just replaces the current instance with a new one inplace (without flushing any file pointers or so on):

> These functions all execute a new program, replacing the current process; they do not return. On Unix, the new executable is loaded into the current process, and will have the same process id as the caller. Errors will be reported as OSError exceptions. The current process is replaced immediately. Open file objects and descriptors are not flushed.

I would suggest changing the _restart function 
[here](https://github.com/Digital-Sapphire/PyUpdater/blob/master/pyupdater/client/updates.py#L250) to something like: 


``
def _restart(self):
    subprocess.Popen((self.current_app, *sys.argv[1:]), stdin=None, stdout=None, stderr=None)
``

Also a  sidenote. When the application is packaged with PyInstaller, it says in the PyInstaller documentation that sys.executable would be equal to the path of the executed bootloader (the main program) of the application (as in, not just the filename).

So wouldn't something like what's done
 [here](https://github.com/Digital-Sapphire/PyUpdater/blob/master/pyupdater/client/updates.py#L250)
 generate some really strange path like "/Applications/AppName.app/Contents/MacOS/Applications/AppName.app/Contents/MacOS/AppName", or does os.path.join take care of that correctly?

**Comment by JMSwag** [comment]
@NicklasTegner I've been unable to reproduce this issue with a few different mac apps that I have. Also, there are test cases for the update process which are passing.

I'd really like to figure out what the root cause is here.

**Comment by JessicaTegner** [comment]
@JMSwag  When restarting the app, does the app actually close for you?
Also. Can't find the test cases for that.

**Comment by JMSwag** [comment]
@NicklasTegner It "looks" as if the previous version closes but I know it's just the new version replacing the old version within the same process. After the update there is only 1 instance of the app running.

[Restart test cases](https://github.com/Digital-Sapphire/PyUpdater/blob/master/tests/test_pyupdater.py#L309)

**Comment by JessicaTegner** [comment]
@JMSwag if I may ask, how complicated are your applications. In my case, the in-process restarting it does on mac, doesn't really seam to work for me.

**Comment by JMSwag** [comment]
@NicklasTegner Not very complicated. Only 1 uses multi pile processes. I'd be willing to merge in a fix that has a switch to restart using system commands(Killing the old version & starting the new version) vs replacing the process.

**Comment by JessicaTegner** [comment]
Nice. I would suggest using subprocess for the task of restarting, also I would make the application in quest responsible for quiting the old (currently running) version.

**Comment by JMSwag** [comment]
Anyone wanna take a stab at the suggestion in #249 but make it an option?

**Comment by JessicaTegner** [comment]
I'll take a look at both suggestions in #249  during the weekend.
